### PR TITLE
Implement @reflectDecl builtin

### PR DESCRIPTION
**Pull request by antlilja**
This PR half implements #10706, it implements the builtin ```@reflectRecl```

### CODE REVIEW

File: lib/std/builtin.zig [hunk 1]
```diff
@@ -183,6 +183,14 @@ pub const SourceLocation = struct {
     column: u32,
 };
 
+/// This data structure is used by the Zig language code generation and
+/// therefore must be kept in sync with the compiler implementation.
+pub const DeclarationInfo = struct {
```
Comment (Vexu) [inline lib/std/builtin.zig:188]:
```suggestion
pub const Declaration = struct {
```
To match `Type` below.

File: src/AstGen.zig [hunk 2]
```diff
@@ -7741,6 +7741,16 @@ fn builtinCall(
             });
             return rvalue(gz, rl, result, node);
         },
+        .reflect_decl => {
+            const container_type = try typeExpr(gz, scope, params[0]);
+            const name = try comptimeExpr(gz, scope, .{ .ty = .const_slice_u8_type }, params[1]);
```
Comment (Vexu) [inline src/AstGen.zig:7746]:
```suggestion
            const name = try comptimeExpr(gz, scope, .{ .coerced_ty = .const_slice_u8_type }, params[1]);
```

File: src/Sema.zig [hunk 3]
```diff
@@ -17974,6 +17975,101 @@ fn zirWasmMemoryGrow(
...
+    const decl_name = try sema.resolveConstString(block, name_src, extra.rhs);
+
+    const target = sema.mod.getTarget();
+
+    try checkNamespaceType(sema, block, ty_src, namespace_ty);
+
+    // Type must be a namespace because we just checked.
+    const namespace = namespace_ty.getNamespace().?;
+
+    if (try sema.lookupInNamespace(block, name_src, namespace, decl_name, true)) |decl_index| {
```
Comment (Vexu) [inline src/Sema.zig:17999]:
```suggestion
    if (try sema.namespaceLookup(block, name_src, namespace, decl_name)) |decl_index| {
```

Is there a specific reason why you copied the check for `is_pub`?

File: src/Sema.zig [hunk 4]
```diff
@@ -17974,6 +17975,101 @@ fn zirWasmMemoryGrow(
...
+            name_src,
+            "declaration of '{s}' could not be found in type '{s}'",
+            .{ decl_name, namespace_name },
+        );
+        errdefer msg.destroy(sema.gpa);
+
+        try sema.mod.errNoteNonLazy(namespace_ty.declSrcLoc(sema.mod), msg, "declared here", .{});
+        break :msg msg;
+    });
+}
```
Comment (Vexu) [inline src/Sema.zig:18072]: Use `failWithBadMemberAccess` preferably with `orelse` to avoid indenting the rest of the function.

**Review by Vexu — CHANGES_REQUESTED** [review]

File: lib/std/builtin.zig (hunk 1 above)
Comment (antlilja) [reply lib/std/builtin.zig:188]: That would conflict with the the already existing Declaration type inside of Type, but maybe Type.Declaration could be renamed to something else but I'm unsure to what that would be.

Comment (Vexu) [reply lib/std/builtin.zig:188]: No need to rename, just use `Type.Declaration` instead of plain `Declaration`.

File: src/Sema.zig (hunk 3 above)
Comment (Vexu) [reply src/Sema.zig:17999]: Be sure to delete the now unnecessary `is_pub` check.

**Review by Vexu — APPROVED** [review]

**Comment by Vexu** [comment]
I noticed that #10706 is no longer accepted so I'm going to close this. It is still linked in the issue and can be easily reopened if it gets accepted later.

**Pull request closed without merging**